File: 567-permutation-in-string/permutation-in-string.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        l1 = len(s1)
        l2 = len(s2)
        d1 = dict(Counter(s1))
        d2 = dict(Counter(s2[:l1]))
        if d1==d2 :
            return True

        for i in range(l1, l2):
            if d1==d2 :
                return True
            
            if s2[i] in d2.keys():
                d2[s2[i]]+=1
            else:
                d2[s2[i]]=1
            
            d2[s2[i-l1]]-=1
            if d2[s2[i-l1]] == 0:
                del d2[s2[i-l1]]

        if d1==d2:
            return True
        
        return False

File: 567-permutation-in-string/permutation-in-string.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        int l1 = s1.length();
        int l2 = s2.length();
        if(l1>l2) return false;
        HashMap<Character, Integer> map1 = new HashMap<>();
        HashMap<Character, Integer> map2 = new HashMap<>();
        for (int i = 0; i < l1; i++) {
            map1.put(s1.charAt(i), map1.getOrDefault(s1.charAt(i), 0) + 1);
        }
        for (int i = 0; i < l1; i++) {
            map2.put(s2.charAt(i), map2.getOrDefault(s2.charAt(i), 0) + 1);
        }
        for (int i = l1; i < l2; i++) {
            if(map1.equals(map2)){
                return true;
            }
            map2.put(s2.charAt(i), map2.getOrDefault(s2.charAt(i), 0) + 1);
            map2.put(s2.charAt(i-l1), map2.get(s2.charAt(i-l1))-1);
            map2.remove(s2.charAt(i-l1), 0);
        }
        if(map1.equals(map2)){
            return true;
        }
        return false;
    }

}